### PR TITLE
Add querybuilder metadata implementation

### PR DESCRIPTION
**Pull request by gwarmstrong**
Depends on #37 

This PR adds the underlying functionality and associated unit tests for processing sample ID queries on metadata, received in the format of [jQuery QueryBuilder](https://querybuilder.js.org/). This will be helpful for adding nice query boxes, e.g below, to a user interface:

Screen Shot 2020-07-17 at 12 25 29 PM

### CODE REVIEW

**Comment by wasade** [comment]
Would it be possible to update this now that #37 has been merged? for some reason the PR isn't picking up that many of these commits are in master

**Comment by gwarmstrong** [comment]
> Would it be possible to update this now that #37 has been merged? for some reason the PR isn't picking up that many of these commits are in master

@wasade  Done.

**Comment by wasade** [comment]
Thanks :) greatly reduced for review now :)

File: microsetta_public_api/api/metadata.py [hunk 1]
```diff
@@ -14,31 +14,33 @@ def category_values(category):
...
+    query = _format_query(kwargs)
+    is_invalid = _validate_query(kwargs, repo)
+    if is_invalid:
+        return is_invalid
+    return _filter_sample_ids(query, repo, alpha_metric, taxonomy)
+
+
 def filter_sample_ids_query_builder(body, taxonomy=None, alpha_metric=None):
-    raise NotImplementedError()
+    query = body
```
Comment (wasade) [inline microsetta_public_api/api/metadata.py:41]: given the API spec, does this actually come in over the signature? I assumed this needed to be obtained off of `request`. But regardless, isn't the `.json` what's desired here?

Comment (wasade) [reply microsetta_public_api/api/metadata.py:41]: also I think the use of `query` in `_filter_sample_ids` is quire a bit different than what the example in the spec shows right?

File: microsetta_public_api/api/tests/implementation_tests/test_metadata.py [hunk 2]
```diff
@@ -145,3 +172,46 @@ def test_metadata_filter_sample_ids_alpha_metric_filter(self):
...
+            response, code = filter_sample_ids_query_builder(
+                self.sample_querybuilder,
+            )
+        self.assertEqual(code, 200)
+        exp = {'sample_ids': ['sample-1', 'sample-3']}
+        obs = json.loads(response)
+        self.assertDictEqual(exp, obs)
+
+    def test_metadata_filter_sample_ids_query_builder_resource_filters(self):
+        with patch.object(MetadataRepo, 'sample_id_matches') as mock_matches, \
```
Comment (wasade) [inline microsetta_public_api/api/tests/implementation_tests/test_metadata.py:192]: I'm not necessarily advocating a change here right now, but I think the use of the mocks may be artificially increasing the complexity of the tests... why not just stage some data and issue a `client.post`?

File: microsetta_public_api/api/metadata.py (hunk 1 above)
Comment (gwarmstrong) [reply microsetta_public_api/api/metadata.py:41]: `query` here is fed directly to `MetadataRepo.sample_id_matches(query)`, which I explicitly designed to accept data that looks like the `QueryBuilder` outputs.

The metadata endpoint we designed, here: https://github.com/biocore/microsetta-public-api/blob/bf4d17307c501a55089c1d0f79425b2964b7c879/microsetta_public_api/api/microsetta_public_api.yml#L77-L96

directly takes the `QueryBuilder`-formatted query as the body of the request, which is exactly what the spec shows in this example:
https://github.com/biocore/microsetta-public-api/blob/bf4d17307c501a55089c1d0f79425b2964b7c879/microsetta_public_api/api/microsetta_public_api.yml#L626-L667

File: microsetta_public_api/api/tests/implementation_tests/test_metadata.py (hunk 2 above)
Comment (gwarmstrong) [reply microsetta_public_api/api/tests/implementation_tests/test_metadata.py:192]: This is the unit test for the implementation layer, which is designed to test the code in `microsetta_public_api.metadata` in isolation. As such, I want to test the code without dependencies on Flask, or or other dependencies.

See the integration tests below for tests that use staged data and test the whole system end-to-end.

File: microsetta_public_api/api/metadata.py (hunk 1 above)
Comment (gwarmstrong) [reply microsetta_public_api/api/metadata.py:41]: You can also consult the API tests for a reference on acceptable inputs for this operation:
https://github.com/biocore/microsetta-public-api/blob/bf4d17307c501a55089c1d0f79425b2964b7c879/microsetta_public_api/api/tests/test_api.py#L246-L356

Comment (wasade) [reply microsetta_public_api/api/metadata.py:41]: I see now... I think flask is just rather permissive in how you interact with body. And ya, mislooked at use with `_filter_sample_ids`

File: microsetta_public_api/api/tests/implementation_tests/test_metadata.py (hunk 2 above)
Comment (wasade) [reply microsetta_public_api/api/tests/implementation_tests/test_metadata.py:192]: Should tests for `microsetta_public_api.metadata` be under that module?

Comment (gwarmstrong) [reply microsetta_public_api/api/tests/implementation_tests/test_metadata.py:192]: I am not sure which tests and which module you are referring to. But generally I will admit the current testing directories structure is a bit wonky.

See #36 for details on the current test structure and a proposed re-organization of the tests in the submodules of the `microsetta_public_api.api` module.

Comment (gwarmstrong) [reply microsetta_public_api/api/tests/implementation_tests/test_metadata.py:192]: Made a fix above that I think should add more clarity.

**Pull request merged**